### PR TITLE
#4 @Intern | Contract-Infra

### PR DESCRIPTION
**Pull request by longtune1802**
Long -intern pull request

### CODE REVIEW

**Comment by longtune1802** [comment]
Please review my last commit 
add test

**Comment by lamsecrettune** [comment]
transferFrom should check allowBalances[from][msg.sender] >= amount, not allowBalances[from][to] >= amount
allowBalances for allowances should be decrease upon transfer successfully

test transfer should check the balance of the sender decrease
test transferFrom should check whether the balance of the source decrease
should have more test like test revert by insufficient balance
lack of testing event emit

**Comment by 0xCheeezzyyyy** [comment]
Inside constructor, a mint function could be used to initially allocate the initial supply to the owner of the contract.

The 'transferFrom' has 2 errors:

1.  In the require statement, should check for msg.sender as spender rather than 'to' i.e. allowBalances[from][msg.sender] >= amount, not allowBalances[from][to] >= amount
2.  I think you forgot to decrement the allowance after transferring on behalf of the user i.e. allowBalances[from][msg.sender] -= amount;

For tests wise, should test for transferFrom with amounts > allowances[owner][spender], this way you could have found out of this bug.

File: contracts/ERC20.sol [hunk 1]
```diff
@@ -29,6 +56,19 @@ contract ERC20 is IERC20 {
...
     uint256 amount
   ) external returns (bool) {
+    require(from!=address(0),"Address should not be 0");
+    require(to!=address(0),"Address should not be 0");
+    // check balance
+    require(balances[from]>=amount,"Insufficient amount");
+    // check allowance
+    require(allowBalances[from][msg.sender]>=amount,"Insufficient allowance");
+    allowBalances[from][msg.sender]-=amount;
+    // decrease value
```
Comment (UncleGrandpa925) [inline contracts/ERC20.sol:71]: you won't need comments for obvious stuff

File: contracts/ERC20.sol [hunk 2]
```diff
@@ -1,26 +1,53 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.8.0;
+pragma solidity ^0.8.11;
 
 import "./IERC20.sol";
 
 contract ERC20 is IERC20 {
+
+  uint private totalBalance;
```
Comment (UncleGrandpa925) [inline contracts/ERC20.sol:8]: Please help implement IERC20Metadata so that we can deploy it on mainnet

Comment (longtune1802) [inline contracts/ERC20.sol:8]: Just forget to push it.
Finish

**Pull request closed without merging**